### ISSUE TITLE
RFC-3: Retiring and replacing the web test runner

### ISSUE THREAD

**Issue by willmorgan** [silverstripe/silverstripe-framework#4254]
RFC-3: Retiring and replacing the web test runner
-----
Status: Draft

---

The core team recently voted in favour of retiring the current web test runner (`TestRunner`) (accessible via `dev/tests`).

The current web test runner functions differently to the sake/phpunit test runners, is susceptible to superglobal leakage, and is basically a bit of a pain to maintain because of the difference in running environments. Also, it's extra code weight and attack surface for live environments when it should really be a dev dependency.

This RFC suggests it is replaced with a module that:
* functions as a web wrapper to call the test runner via the `phpunit` binary
* with a [listener](https://phpunit.de/manual/4.7/en/extending-phpunit.html#extending-phpunit.PHPUnit_Framework_TestListener) feeding back to the UI via WebSocket.
* would run only in dev mode
* strongly discourage being loaded as anything but a dev dependency

First steps required:
* community scrutiny
* a proof of concept :wink:

Ingo's initial post on this: https://groups.google.com/forum/#!topic/silverstripe-dev/I5dHy9dMVng

**Comment by micmania1** [comment]
I would argue that it doesn't need to be put into a module - does anybody really want to maintain something we don't need?

**Comment by chillu** [comment]
Hey @willmorgan, do you want to incorporate my post into this RFC? https://groups.google.com/forum/#!topic/silverstripe-dev/I5dHy9dMVng
For example, I had already outlined what the equivalent replacement for each `sake` command would be, detail which is important for an RFC. 

I think while a `phpunit` web wrapper and websocket implementation would be neat, its not required to get the functionality out of core (my main goal). I suggest we put `TestRunner` into a module as-is, mark it as unmaintained, and ask for maintainers on the mailing list. That allows users who rely on this feature to continue using it, but doesn't take up core maintainers time with maintenance (my secondary goal). My underlying assumption here is that its less work to just separate out the functionality than to create this new wrapper. Will, what do you think?

At the core discussion, we also raised that docs need to be "beginner friendly" to make CLI usage as a default feasible. I think we're doing fairly well on http://docs.silverstripe.org/en/3.1/developer_guides/testing/ already. We do need to mention the `$_FILE_TO_URL_MAPPING` setting though, its only explained on http://docs.silverstripe.org/en/3.1/developer_guides/cli/ and http://docs.silverstripe.org/en/3.1/getting_started/environment_management/. At the moment, you can't even run `sake dev/build` on CLI without setting that, because `SiteTree->requireDefaultRecords()` calls `Director::protocolAndHost()`.

On the long run, I think we should investigate making `$_FILE_TO_URL_MAPPING` optional, which essentially means rewriting the few `Director` functions currently relying on `Director::protocolAndHost()`. But that's a different RFC ;)

**Comment by tractorcow** [comment]
I think putting it into a module is a great way of helping users who rely on features we don't want to maintain anymore.

**Comment by willmorgan** [comment]
> My underlying assumption here is that its less work to just separate out the functionality than to create this new wrapper. Will, what do you think?

Mostly agreed. I think @dhensby and I see more utility in running the web test runner, and the main problem for us is the inconsistency in test environments, so a wrapper would be useful for us.

I think in the first instance, separating out the `TestRunner` in to a module as-is would be a great first step.

**Comment by simonwelsh** [comment]
> I think putting it into a module is a great way of helping users who rely on features we don't want to maintain anymore.

If you put it in a module, you still need to maintain it.

**Comment by dhensby** [comment]
* actively develop

FTFY

**Comment by chillu** [comment]
> If you put it in a module, you still need to maintain it.

I suggest that we'll deprecate it within our usual deprecation cycle for core use and maintain it during this cycle ends (until 5.0).

**Comment by camfindlay** [comment]
RFC-3 (Just a note this one isn't in the format proposed, however that hasn't been accepted into docs yet so I'll let you away with it).

**Comment by tractorcow** [comment]
We don't need to maintain or actively develop any module outside of core + the list of modules we are contractually obliged to support, plus any dependencies it requires. For example, siteconfig and reports have been modularised, but they are still core dependencies since CMS relies on these directly, and thus are supported. Translatable was once in core, but was removed in 3.0, and thus isn't considered a part of core anymore.

If we decide we wish to remove support and/or deprecate for a particular feature, but decide to put it in an external module, then we can't do that outside of a major release version. I wouldn't imagine this would be even a require-dev, so I can't see why we would be providing support for it. The earliest this could happen is 4.0.

What we DO need to support is any version of the core module, up until the end of life, and this includes version of 3.x which contain the testrunner within core.

Saying "we will support this feature even though it's no longer a part of framework and is deprecated" is a risky venture, and we won't do it unless we have good reason to do so. We will support it as a part of any 3.x release but likely not beyond that.

If a 4.x compatible testrunner module exists, then it's really up to whether we have time or the goodwill to maintain it, or otherwise if someone in the community wishes to take on that task.

**Comment by simonwelsh** [comment]
That attitude from SS Ltd is actively toxic to the module ecosystem. The amount of "release and forget" modules under both silverstripe and silverstripe-labs is massive and is a really bad sign of what's expected of modules. Bit rot is a very real problem and SS is one of the main sources of that problem.

**Comment by edlinklater** [comment]
> We don't need to maintain or actively develop any module outside of core + the list of modules we are contractually obliged to support

This is a bit confusing - should probably clarify that by "we" here you mean SS Ltd and not the open source project / core committers. The project/core has no contractual obligation to support anything, and SS Ltd doesn't decide what is and isn't supported by project/core.

**Comment by dhensby** [comment]
I'm new to this SS thing, but I believe SS are contractually obliged to support some modules as part of CWP

**Comment by tractorcow** [comment]
By "we" I mean SilverStripe, but you're right, we don't forbid other contributors (or community members) from offering support for modules. We will, however, not give the impression that a certain module is officially supported by SilverStripe when that's not the case.

> I'm new to this SS thing, but I believe SS are contractually obliged to support some modules as part of CWP

Yes that's right, and we are looking at coming up with a better mechanism for helping identify modules as "supported" and clarifying what "supported" means.

**Comment by camfindlay** [comment]
And by "SilverStripe" @tractorcow you mean SilverStripe Ltd.

**Comment by tractorcow** [comment]
Thanks for that, updated my reply :D

**Comment by SpiritLevel** [comment]
...tricky balancing act between Ltd's and community's interests...and I don't just mean remembering to include or exclude the "Ltd" ;)

**Comment by SpiritLevel** [comment]
@simonwelsh.  I think it would be more toxic to withhold code, eliminating the possibility that someone else may take it up and develop it further.

**Comment by jonom** [comment]
The SilverStripe brand name is quite confusing... it would be nice if the company and the open source project had different names. Like Ellis Labs has CodeIgniter and Expression Engine.

**Comment by willmorgan** [comment]
> If you don't want to support something don't release it. If you do release it, support it. It's a fairly simple concept.

That's a bold viewpoint to have. I assume by "support" you mean offer a quick response and resolution to all issues and feature requests filed? :wink:

Simon, as I've no doubt you know, SS's open source core, and most of its modules, use the [BSD license](https://en.wikipedia.org/wiki/BSD_licenses#3-clause_license_.28.22Revised_BSD_License.22.2C_.22New_BSD_License.22.2C_or_.22Modified_BSD_License.22.29); it clearly sets out commitment to warranty. In practice everyone is more diligent, but ultimately if you're offering code for free, often as a byproduct of client work that generates you actual revenue, I think it's fair to expect that other users can and should muck in to help.

In my personal opinion, the open source SS project already does too much and reinvents the wheel in places that aren't necessary.

The spirit of this RFC is to encourage usage of the proper PHPUnit binary, reduce the codebase, and provide a lightweight fallback for the CLI-shy.

**Comment by dhensby** [comment]
@simon_w I just picked a random project from your public gitlab and checked the licence: 

>THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 

Your project explicitly rule out any kind of warranty to maintain it but you seem to be happy to say we should be giving explicit warranty. Why is it that you don't provide a warranty if you feel it is an essential aspect of releasing OSS?

**Comment by SpiritLevel** [comment]
Ouch.

**Comment by camfindlay** [comment]
Guys, please keep the discussion focused on the RFC at hand. If you want to discuss the in and outs of open source licensing and shared maintenance of FOSS please find an appropriate discussion space for this.

**Comment by simonwelsh** [comment]
Looks like I'm being stifled again for having a relevant discussion.

@willmorgan I'm fine with the support being removed so long as if SS Ltd does release it as a module, they support it.

Since my opinions aren't wanted, that's the last I'll be saying on this.

**Comment by sminnee** [comment]
I don't think we should release this as a module—it's not really relevant to an SS4 user, and the bitrot is likely to cause more harm than good.

In other areas (e.g. legacydatetimefields), we've release features that we've dropped as modules because it's likely that it will be frustrating for people to upgrade unless we do that. We've also put "legacy" in the name that suggests that maybe people shouldn't use it for new projects. ;-) We can probably do better, e.g. tagging the module as "unmaintained" and excluding modules tagged this way from addons results.

However, I can't imagine that dropped the web test runner is going to cause upgrade difficulties.

If someone complains / asks for it to be available as a module, we can give them some pointers for how we can make it a module should they want to maintain it themselves.

**Comment by chillu** [comment]
We've removed the wrappers from master through https://github.com/silverstripe/silverstripe-framework/pull/4404. I share @sminnee and @micmania's viewpoints, and don't think a module is required for this. In 90% of the test runs performed through `dev/tests`, you have an equivalent phpunit command. If somebody cares strongly about running tests through a browser, they're welcome to start a module.

Closing this RFC as "partially done", with the actual wrappers removed from `master` now. Since a module for `dev/tests` doesn't have any core dependency, it doesn't need to go through an RFC in core.
